src/modules/iotjs_module_mqtt.c

```diff
@@ -0,0 +1,677 @@
+/* Copyright 2018-present Samsung Electronics Co., Ltd. and other contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <stdlib.h>
+#include <string.h>
+
+#include "iotjs_def.h"
+#include "iotjs_module_buffer.h"
+#include "iotjs_module_mqtt.h"
+
+
+#include "iotjs_handlewrap.h"
+#include "iotjs_reqwrap.h"
+
+static void iotjs_mqtt_client_destroy(iotjs_mqtt_client_t *mqtt_client) {
+  IOTJS_RELEASE(mqtt_client);
+}
+
+static const jerry_object_native_info_t mqtt_client_native_info = {
+  .free_cb = (jerry_object_native_free_callback_t)iotjs_mqtt_client_destroy
+};
+
+static iotjs_mqtt_client_t *iotjs_mqtt_client_create(
+    const jerry_value_t jobject) {
+  iotjs_mqtt_client_t *mqtt_client = IOTJS_ALLOC(iotjs_mqtt_client_t);
+
+  jerry_set_object_native_pointer(jobject, mqtt_client,
+                                  &mqtt_client_native_info);
+
+  return mqtt_client;
+}
+
+
+IOTJS_DEFINE_NATIVE_HANDLE_INFO_THIS_MODULE(mqtt_client);
+
+
+static jerry_value_t mqtt_client_connack_error(const unsigned char error_code) {
+  switch (error_code) {
+    case UNACCEPTABLE_PROTOCOL:
+      return JS_CREATE_ERROR(COMMON,
+                             "MQTT: Connection refused: unacceptable protocol");
+    case BAD_IDENTIFIER:
+      return JS_CREATE_ERROR(COMMON,
+                             "MQTT: Connection refused: bad client identifier");
+    case SERVER_UNAVIABLE:
+      return JS_CREATE_ERROR(COMMON,
+                             "MQTT: Connection refused: server unaviable");
+    case BAD_CREDENTIALS:
+      return JS_CREATE_ERROR(
+          COMMON, "MQTT: Connection refused: bad username or password");
+    case UNAUTHORISED:
+      return JS_CREATE_ERROR(COMMON, "MQTT: Connection refused: unauthorised");
+    default:
+      return JS_CREATE_ERROR(COMMON, "MQTT: Unknown error");
+  }
+}
+
+
+/*
+ * Waits for a uint32_t remaining_length.
+ * Returns the encoded value with the correct size.
+ */
+static char *iotjs_encode_remaining_length(uint32_t len) {
```

Comment:
I feel this is overcomplicated. You could write the data directly into the stream.

Reply:
We could, but we need the size of that data to allocate the buffer.

Reply:
I figured out a solution for this.